syslog_rfc5424_parser/parser.py

```diff
@@ -1,64 +1,132 @@
-    msgid.setResultsName('msgid')
-)
-
-sd_name = Word(NameAscii, min=1, max=32)
-sd_id = sd_name
-param_name = sd_name
-sd_param = Group(
-    param_name.setResultsName('param_name') + Suppress(Literal("=")) +
-    QuotedString(quoteChar='"', escChar='\\', escQuote='\\').setResultsName('param_value')
-)
-
-sd_element = Group(
-    Suppress("[") + sd_id.setResultsName('sd_id') + ZeroOrMore(SP + sd_param).setResultsName('sd_params') +
-    Suppress("]")
-)
-
-structured_data = (NilValue | OneOrMore(sd_element)).setResultsName('sd_element')
-
-msg = Combine(restOfLine + lineEnd)
-
-syslog_message = header.setResultsName('header') + SP + structured_data.setResultsName('sd')
-
-
-__all__ = ['syslog_message', 'structured_data', 'header']
+import collections
+
+from lark import Lark, Transformer
+
+
+GRAMMAR = '''
+    ?start           : header _SP structured_data [ msg ]
+    ?header          : pri version _SP timestamp _SP hostname _SP appname _SP procid _SP msgid
+    pri              : "<" /[0-9]{1,3}/ ">"
+    version          : /[1-9][0-9]{0,2}/
+    timestamp        : NILVALUE
+                     | /[0-9]{4}-[0-9]{2}-[0-9]{2}T[0-9]{2}:[0-9]{2}:[0-9]{2}/ time_secfrac? time_offset
+    time_secfrac     : /\.[0-9]{1,6}/
+    time_offset      : ZULU
+                     | _ntime_offset
+    _ntime_offset    : /[+-][0-9]{2}:[0-9]{2}/
+    structured_data  : NILVALUE
+                     | sd_element+
+    sd_element       : "[" sd_id (" " sd_param)* "]"
+    ?sd_id           : sd_name
+    sd_param         : param_name "=" ESCAPED_STRING
+    ?param_name      : sd_name
+    ?sd_name         : /[^= \]\"]{1,32}/
+    appname          : NILVALUE
+                     | /[!-~]{1,48}/
+    procid           : NILVALUE
+                     | /[!-~]{1,128}/
+    msgid            : NILVALUE
+                     | /[!-~]{1,32}/
+    hostname         : NILVALUE
+                     | /[!-~]{1,255}/
+    msg              : / .*/ms
+
+    %import common.ESCAPED_STRING   -> ESCAPED_STRING
+
+    _SP: " "
+    NILVALUE: "-"
+    ZULU: "Z"
+'''
+
+
+Header = collections.namedtuple('Header', ['pri', 'version', 'timestamp', 'hostname', 'appname', 'procid', 'msgid'])
+
+SDElement = collections.namedtuple('SDElement', ['sd_id', 'sd_params'])
+
+ParsedMessage = collections.namedtuple('ParsedMessage', ['header', 'structured_data', 'message'])
+
+
+class TreeTransformer(Transformer):
+    def NILVALUE(self, inp):
+        return '-'
+
+    def pri(self, inp):
+        return int(inp[0])
+
+    def version(self, inp):
+        return int(inp[0])
+
+    def timestamp(self, inp):
+        if len(inp) == 1:
+            return inp[0]
+        else:
+            datetime = str(inp[0])
+            rest = [str(i.children[0]) for i in inp[1:]]
+            return datetime + ''.join(rest)
+
+    def hostname(self, inp):
+        return str(inp[0])
+
+    def appname(self, inp):
+        return str(inp[0])
+
+    def procid(self, inp):
```

Comment:
I don't remember if it's cheaper to convert an int to a string back to an int or to test if it's an int initially.